### PR TITLE
Add security to Riak

### PR DESCRIPTION
**Pull request by Vagabond**
This PR implements all the riak_core bits of security, that is, a place security information is stored and modified and APIs for checking a user has a particular set of permissions. riak_core applications are expected to use these APIs themselves, which is why there's no actual security checking done anywhere in this PR. The riak_kv and riak_api repos are where that happens.

See also basho/riak#355

### CODE REVIEW

File: src/table.erl [hunk 1]
```diff
@@ -0,0 +1,84 @@
+-module(table).
```
Comment (engelsanchez) [inline src/table.erl:1]: So this is really the definitive table in the Erlang world? :)

Comment (engelsanchez) [inline src/table.erl:1]: Licence header or it didn't happen

File: src/table.erl [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    lists:reverse([BottomBorder | IoList]);
+create_table(_Spec, [], _Length, IoList) ->
+    lists:reverse(IoList);
+create_table(Spec, [Row | Rows], Length, IoList) ->
+    create_table(Spec, Rows, Length, [row(Spec, Row) | IoList]).
+
+-spec get_row_length(list(tuple())) -> non_neg_integer().
+get_row_length(Spec) ->
+    lists:foldl(fun({_Name, Size}, Total) ->
+                    Total + Size + 2
```
Comment (engelsanchez) [inline src/table.erl:33]: I think 7 is a prettier number

File: rebar.config [hunk 3]
```diff
@@ -12,5 +12,6 @@
   {basho_stats, ".*", {git, "git://github.com/basho/basho_stats", {branch, "develop"}}},
   {riak_sysmon, ".*", {git, "git://github.com/basho/riak_sysmon", {branch, "develop"}}},
   {folsom, ".*", {git, "git://github.com/basho/folsom.git", {tag, "0.7.4p2"}}},
-  {ranch, "0.4.0-p1", {git, "git://github.com/basho/ranch.git", {tag, "0.4.0-p1"}}}
-       ]}.
+  {ranch, "0.4.0-p1", {git, "git://github.com/basho/ranch.git", {tag, "0.4.0-p1"}}},
+  {pbkdf2, ".*", {git, "git://github.com/basho/erlang-pbkdf2", {branch, "adt-cleanups"}}}
```
Comment (jaredmorrow) [inline rebar.config:16]: Mental note that we should merge adt-cleanups to master, give pdkdf2 the 'develop' treatment and then update this line.

File: src/riak_core.erl [hunk 4]
```diff
@@ -350,10 +350,15 @@ register(App, [{bucket_validator, ValidationMod}|T]) ->
...
+register(App, [{permissions, Permissions}|T]) ->
+    register_mod(App, Permissions, permissions),
+    register(App, T);
+register(App, [{auth_mod, {AuthType, AuthMod}}|T]) ->
+    register_proplist({AuthType, AuthMod}, auth_mods),
     register(App, T).
 
-
-register_mod(App, Module, Type) when is_atom(Module), is_atom(Type) ->
+register_mod(App, Module, Type) when is_atom(Type) ->
```
Comment (jaredmorrow) [inline src/riak_core.erl:361]: Why this change?

Comment (Vagabond) [inline src/riak_core.erl:361]: Because when we register permissions, it is a list. not an atom.

File: src/riak_core_security.erl [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+
+-record(context,
+        {username,
+         grants,
+         epoch}).
+
+prettyprint_users([all]) ->
+    "all";
+prettyprint_users(Users0) ->
+    %% my kingdom for an iolist join...
```
Comment (jaredmorrow) [inline src/riak_core_security.erl:39]: R17 PR's accepted

**Comment by jaredmorrow** [comment]
Anything else need to be done now that we decided on R16 with conditional support for R15?

**Comment by jaredmorrow** [comment]
+1 from testing on the command line the functionality and code reading with the understanding that there are some `%TODO` we need to get rid of before this hits staging.

File: src/table.erl (hunk 1 above)
Comment (Vagabond) [inline src/table.erl:1]: This will be cleaned up and renamed for pre3.
